src/communicate.rs

```diff
@@ -145,96 +145,110 @@ pub fn write_to_peer_connection(
+    let leaf = async move {
+        let mut o_stream = match uni_stream.await {
+            Ok(o_stream) => o_stream,
+            Err(e) => {
                 utils::handle_communication_err(
                     peer_addr,
                     &From::from(e),
-                    "Shutdown-after-write",
-                    user_msg1,
-                )
-            })
-        })
-        .map(move |_| utils::handle_send_success(peer_addr, user_msg2));
+                    "Open-Unidirectional",
+                    user_msg,
+                );
+                return;
+            }
+        };
 
-    current_thread::spawn(leaf);
-}
+        let (message, msg_flag) = wire_msg.into();
 
-/// Listen for incoming streams containing peer messages and read them when available
-pub fn read_from_peer(peer_addr: SocketAddr, incoming_streams: quinn::IncomingStreams) {
-    let leaf = incoming_streams
-        .map_err(move |e| {
+        if let Err(e) = o_stream.write_all(&message[..]).await {
+            utils::handle_communication_err(peer_addr, &From::from(e), "Write-All", user_msg);
+            return;
+        }
+        if let Err(e) = o_stream.write_all(&[msg_flag]).await {
+            utils::handle_communication_err(peer_addr, &From::from(e), "Write-All", user_msg);
+            return;
+        }
+
+        if let Err(e) = o_stream.finish().await {
             utils::handle_communication_err(
                 peer_addr,
                 &From::from(e),
-                "Incoming streams failed",
-                None,
+                "Shutdown-after-write",
+                user_msg,
             );
-        })
-        .for_each(move |quic_stream| {
-            read_peer_stream(peer_addr, quic_stream).map_err(|e| {
-                debug!(
-                    "Error in Incoming-streams while reading from peer {}: {:?} - {}.",
-                    peer_addr, e, e
-                )
-            })
-        });
+            return;
+        }
+
+        utils::handle_send_success(peer_addr, user_msg);
+    };
 
-    current_thread::spawn(leaf);
+    let _ = tokio::spawn(leaf);
 }
 
-fn read_peer_stream(peer_addr: SocketAddr, quic_stream: quinn::NewStream) -> R<()> {
-    let i_stream = match quic_stream {
-        quinn::NewStream::Bi(_, _) => {
-            let e = Error::BiDirectionalStreamAttempted(peer_addr);
-            utils::handle_communication_err(peer_addr, &e, "Receiving Stream", None);
-            return Err(e);
+/// Listen for incoming streams containing peer messages and read them when available
+pub fn read_from_peer(
+    peer_addr: SocketAddr,
+    mut uni_streams: quinn::IncomingUniStreams,
+    mut bi_streams: quinn::IncomingBiStreams,
+) {
+    let _ = tokio::spawn(async move {
+        if let Some(res) = bi_streams.next().await {
+            let err = match res {
+                Err(e) => {
+                    debug!(
+			"Error in Incoming-bi-stream while reading from peer {}: {:?} - {}.\nNote: It
```

Comment:
Although bi-directional are allowed for clients? Perhaps reword to make that clear?

Reply:
i.e. p2p nodes cannot be connected bi-directionally. Client connections are bi-directional, but p2p nodes cannot be.

Reply:
>Although bi-directional are allowed for clients?

Not quite, it's a bit confusing thing, but we don't use bi-directional streams at all. For clients it's more like 2 uni-directional streams instead of 1 bi-directional.